[AArch64] Extend condition optimizer to support unsigned comparisons

AZero13 opened the pull request:
We have to be extra careful to not allow unsigned wraps, however. This also required some adjusting of the logic in adjustCmp, as well as compare the true imm value with add or sub taken into effect.

Because SIGNED_MIN and SIGNED_MAX cannot be an immediate, we do not need to worry about those edge cases when dealing with unsigned comparisons.

AZero13 commented:
@davemgreen Do you know where LLVM actually removes the redundant instructions?

AZero13 commented:
Future PR: I don't want to have a situation where we jump via bls -> blo -> where we wanna go

Instead we can just do 

blo (if bls is first) blo-loc
then:
bls bls-loc because one implies the other.

AZero13 commented:
@paulwalker-arm Can I have your opinion on this?

AZero13 commented:
@davemgreen Thoughts on this?

davemgreen commented on llvm/test/CodeGen/AArch64/combine-comparisons-by-cse.ll:
```diff
@@ -845,6 +845,668 @@ return:                                           ; preds = %if.end, %land.lhs.t
   ret i32 %retval.0
 }
 
+; (a > 10 && b == c) || (a >= 10 && b == d)
```
Is it possible to make simpler tests too? Even if they are mir tests.
Some of the issues in these can be very subtle and easy to get wrong. (And they can make my head hurt). Whatever we can do to keep things simple and clear, the better, and thorough tests can help with that. I got a report of a potential problem with #143013 but that appears to be caused by a different issue, not anything related to #143013.

AZero13 replied:
I just copy pasted the original tests and changed signed to unsigned.

AZero13 commented:
@davemgreen I tried to make the tests. However, I found it best to just copy paste the original tests, changed signed to unsigned, even variants that mixed both signed and unsigned.

AZero13 commented:
So, basically everything was tested. Which is why I made sure to ensure that all cases were covered.

For example: x > 0 || x < -1, does not work when unsigned, so how do I check this? by checking that we switched from adds to subs. 

I used a helper function to check between both signed and unsigned variants, and the realheadimm was because it is important to know what we are truly comparing again.

Now, it is not obvious how this actually relates to https://github.com/llvm/llvm-project/pull/143013. 

That is my fault, but honestly, fixing the root cause of one of the issues responsible for that is required to get this to work too.

How? Well, look at adjustCmp. It was flipping the sign at cmn 0 to cmp 1, when it should have been cmp 0. And therefore, the logic was flawed to begin with. And this is important as when we do a signed wrap, we change from subs to adds, etc.

topperc commented on llvm/lib/Target/AArch64/AArch64ConditionOptimizer.cpp:
```diff
@@ -227,26 +227,43 @@ static int getComplementOpc(int Opc) {
   unsigned Opc = CmpMI->getOpcode();
+  unsigned OldOpc = Opc;
+
+  bool isSigned = Cmp == AArch64CC::GT || Cmp == AArch64CC::GE ||
```
Capitalize `IsSigned` per coding standards

AZero13 replied:
Done!

davemgreen reviewed (commented):
Why can't we write simpler mir tests for this?

AZero13 commented:
> These are literally just copy pastes of the old tests but modified to work with unsigned and mixed versions @davemgreen
> 
> There's nothing that complex about it.

To prevent the conditions from being condensed to a conditional compare, it needs to be complex to ensure these conditions are implemented as branches in the asm.

davemgreen commented:
Hi. Ideally we would be able to come up with a formal proof to ensure we are not missing anything. Like you said some of the cases are difficult to reach though. Something like this is what I was thinking of for  mir test, that can be more concise but we can still use for testing. Different variants can be made up.
```
# NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
# RUN: llc -mtriple=aarch64 -o - %s -run-pass=aarch64-condopt | FileCheck %s

---
name:            add_add_hi_hi
tracksRegLiveness: true
body:             |
  ; CHECK-LABEL: name: add_add_hi_hi
  ; CHECK: bb.0:
  ; CHECK-NEXT:   successors: %bb.1(0x40000000), %bb.2(0x40000000)
  ; CHECK-NEXT:   liveins: $w0
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:gpr32common = COPY $w0
  ; CHECK-NEXT:   [[ADDSWri:%[0-9]+]]:gpr32 = ADDSWri [[COPY]], 0, 0, implicit-def $nzcv
  ; CHECK-NEXT:   Bcc 12, %bb.1, implicit $nzcv
  ; CHECK-NEXT:   B %bb.2
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT: bb.1:
  ; CHECK-NEXT:   successors: %bb.2(0x80000000)
  ; CHECK-NEXT:   liveins: $w0
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT:   [[ADDSWri1:%[0-9]+]]:gpr32 = ADDSWri [[COPY]], 1, 0, implicit-def $nzcv
  ; CHECK-NEXT:   Bcc 12, %bb.2, implicit $nzcv
  ; CHECK-NEXT:   B %bb.2
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT: bb.2:
  ; CHECK-NEXT:   liveins: $w0
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT:   [[MOVi32imm:%[0-9]+]]:gpr32 = MOVi32imm 1
  ; CHECK-NEXT:   $w0 = COPY [[MOVi32imm]]
  ; CHECK-NEXT:   RET undef $lr, implicit $w0
  bb.0:
    liveins: $w0
    %0:gpr32common = COPY $w0
    %4:gpr32 = ADDSWri %0:gpr32common, 0, 0, implicit-def $nzcv
    Bcc 12, %bb.2, implicit $nzcv
    B %bb.3

  bb.2:
    liveins: $w0
    %5:gpr32 = ADDSWri %0:gpr32common, 1, 0, implicit-def $nzcv
    Bcc 12, %bb.3, implicit $nzcv
    B %bb.3

  bb.3:
    liveins: $w0
    %1:gpr32 = MOVi32imm 1
    $w0 = COPY %1
    RET undef $lr, implicit $w0
...
```

davemgreen commented on llvm/lib/Target/AArch64/AArch64ConditionOptimizer.cpp:
```diff
@@ -248,13 +265,23 @@ AArch64ConditionOptimizer::CmpInfo AArch64ConditionOptimizer::adjustCmp(
-  if (OldImm == 0 && ((Negative && Correction == 1) ||
-                      (!Negative && Correction == -1))) {
+  // Handle cmn 1 -> cmp 0, transitions by adjusting compare instruction opcode.
+  if (OldImm == 1 && Negative && Correction == -1) {
```
Can you give more details why is this valid? It doesn't seem valid in isolation, but you assert for ADDS 0 below anyway so there wouldn't be anything to CSE.
I don't think there is a testcase for it.

SiliconA-Z commented on llvm/lib/Target/AArch64/AArch64ConditionOptimizer.cpp:
```diff
@@ -248,13 +265,23 @@ AArch64ConditionOptimizer::CmpInfo AArch64ConditionOptimizer::adjustCmp(
+  }
+
+  // Handle +0 -> -1 transitions by adjusting compare instruction opcode.
+  assert((OldImm != 0 || !Negative) && "Should not encounter cmn 0!");
```
adds 0 should never happen.

subs 0 always sets the flag correctly.

Adds 0 always resets the flags

davemgreen replied:
If the instruction is valid then the pass should handle it correctly. If it isn't valid then it should be a verifier error, but I don't believe there is anything invalid about it. Passes should be able to handle all valid IR correctly, as we don't know how things will change in the future.

davemgreen reviewed (commented):
This is still asserting on valid instructions AFAICT (even if they are unlikely to be generated in practice at the moment). That needs to be fixed first, probably just by bailing out if it sees one?

It otherwise seems to be a sensible patch and looks like it comes up enough to make it worthwhile.

AZero13 committed:
[AArch64] Add support for unsigned comparisons

We have to be extra careful to not allow unsigned wraps, however. This also required some adjusting of the logic in adjustCmp, as well as compare the true imm value with add or sub taken into effect.

Because SIGNED_MIN and SIGNED_MAX cannot be an immediate, we do not need to worry about those edge cases when dealing with unsigned comparisons.

davemgreen reviewed (approved):
I ran a lot of tests and they look to be doing OK. This LGTM but I have a suggestion on making it slightly more straight forward.

AZero13 committed:
Address comments
